rsconnect/api.py

```diff
@@ -819,53 +819,56 @@ def save_deployed_info(self, *args, **kwargs):
 
     @cls_logged("Validating app mode...")
     def validate_app_mode(self, *args, **kwargs):
-        path = (
-            self.get("path", **kwargs)
-            or self.get("file", **kwargs)
-            or self.get("file_name", **kwargs)
-            or self.get("directory", **kwargs)
-            or self.get("file_or_directory", **kwargs)
-        )
-        app_store = self.get("app_store", *args, **kwargs)
-        if not app_store:
-            module_file = fake_module_file_from_directory(path)
-            self.state["app_store"] = app_store = AppStore(module_file)
-        new = self.get("new", **kwargs)
-        app_id = self.get("app_id", **kwargs)
-        app_mode = self.get("app_mode", **kwargs)
-
-        if new and app_id:
-            raise RSConnectException("Specify either a new deploy or an app ID but not both.")
-
-        existing_app_mode = None
-        if not new:
-            if app_id is None:
-                # Possible redeployment - check for saved metadata.
-                # Use the saved app information unless overridden by the user.
-                app_id, existing_app_mode = app_store.resolve(self.remote_server.url, app_id, app_mode)
-                logger.debug("Using app mode from app %s: %s" % (app_id, app_mode))
-            elif app_id is not None:
-                # Don't read app metadata if app-id is specified. Instead, we need
-                # to get this from the remote.
-                if isinstance(self.remote_server, RSConnectServer):
-                    app = get_app_info(self.remote_server, app_id)
-                    existing_app_mode = AppModes.get_by_ordinal(app.get("app_mode", 0), True)
-                elif isinstance(self.remote_server, PositServer):
-                    app = get_rstudio_app_info(self.remote_server, app_id)
-                    existing_app_mode = AppModes.get_by_cloud_name(app.json_data["mode"])
-                else:
-                    raise RSConnectException("Unable to infer Connect client.")
-            if existing_app_mode and existing_app_mode not in (None, AppModes.UNKNOWN, app_mode):
-                msg = (
-                    "Deploying with mode '%s',\n"
-                    + "but the existing deployment has mode '%s'.\n"
-                    + "Use the --new option to create a new deployment of the desired type."
-                ) % (app_mode.desc(), existing_app_mode.desc())
-                raise RSConnectException(msg)
-
-        self.state["app_id"] = app_id
-        self.state["app_mode"] = app_mode
-        return self
+        try:
```

Comment:
Marking this entire block for one error feels a bit odd, as other unrelated errors might occur. Another way to do this could be to use get_app_info to check whether the app still exists. The place for adding the specific exception would be L847-L857.

Reply:
Thanks, @bcwu - I've updated the try/except block to be more specific.